### PR TITLE
[ASImageProtocols] Fix nullables to avoid bridging crashes

### PR DESCRIPTION
**Pull request by flovouin**
Hi guys,

Following #3112 I'm submitting this PR to solve the bridging crashes I was experiencing when implementing `ASImageCacheProtocol` in Swift.
A sample project showcasing the problem can be found here: https://github.com/flovouin/ASDK-ASImageCacheProtocolBridge
The problem is that `clearFetchedImageFromCacheWithURL` is sometimes called with a `nil` URL when it shouldn't.

So I double checked all the calls to `ASImageCacheProtocol` and `ASImageDownloaderProtocol` in the ASDK codebase and came up with the following PR.

- The first commit ensures that the image nodes never call the image protocols methods with a `nil` URL / identifier when they shouldn't. The only place where I am not entirely sure whether the URL can be `nil` is in the [ASNetworkImageNode](https://github.com/flovouin/AsyncDisplayKit/blob/d1e7759e7835521b2b1b9b00a9fae7bea7a70185/Source/ASNetworkImageNode.mm#L505). It would seem that `_downloadImageWithCompletion` is only called if `_URL != nil`, but as there's some multi thread funkyness happening, I'm not sure it is safe to assume. If someone who has more experience than me could review this part, that'd be nice :-)
- The second commit removes unnecessary checks for `nil` identifiers in image downloaders, as they are declared as `nonnull` in the protocol definitions.

@maicki was suggesting in #3112 to have `synchronouslyFetchedCachedImageWithURL` accept a `nullable` URL, which is actually already the case, although I'm not sure it really makes sense: why would you ask for the resource corresponding to a `nil` URL, especially if you already know the result? Also, this method is currently only called with a (explicitly checked) non-nil URL in the codebase. Finally, its documentation seems outdated as it references a `completion` block that does not exist in the method prototype. However I did not amend it for the time being, please tell me if you want me to do so :-) I personally would be in favour of making the URL argument `nonnull`.

This PR solves the crashes demonstrated by the sample project, but I'm willing to make more changes if you have any remarks!

Cheers,
Flo

### CODE REVIEW

File: Source/ASMultiplexImageNode.mm [hunk 1]
```diff
@@ -203,7 +203,10 @@ - (void)didExitPreloadState
   [self _setDownloadIdentifier:nil];
   
   if (_cacheSupportsClearing && self.loadedImageIdentifier != nil) {
-    [_cache clearFetchedImageFromCacheWithURL:[_dataSource multiplexImageNode:self URLForImageIdentifier:self.loadedImageIdentifier]];
+    NSURL *URL = [_dataSource multiplexImageNode:self URLForImageIdentifier:self.loadedImageIdentifier];
+    if (URL) {
```
Comment (maicki) [inline Source/ASMultiplexImageNode.mm]: Please use `URL != nil`

File: Source/ASNetworkImageNode.mm [hunk 2]
```diff
@@ -527,7 +527,9 @@ - (void)_downloadImageWithCompletion:(void (^)(id <ASImageContainerProtocol> ima
     }
     
     if (cancelAndReattempt) {
-      [_downloader cancelImageDownloadForIdentifier:downloadIdentifier];
+      if (downloadIdentifier) {
```
Comment (maicki) [inline Source/ASNetworkImageNode.mm]: Please use `downloadIdentifier != nil`

**Comment by flovouin** [comment]
Thanks for the review @maicki ! There, I made the requested changes. If there's anything else I can do, just ask :-)

**Comment by flovouin** [comment]
Thanks guys!

**Pull request merged**